### PR TITLE
Faster implementation for VMC_SRt

### PR DESCRIPTION
**Pull request by riccardo-rende**
This implementation is based on neural tangents, and the new driver is called VMC_SRt_ntk.
cc: @llviteritti @alleSini99 @gcarleo

### CODE REVIEW

File: netket/experimental/driver/vmc_srt_ntk.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+
+from netket.utils.optional_deps import import_optional_dependency
+
+from einops import rearrange
+from netket.jax.sharding import shard_along_axis, all_gather
+
+nt = import_optional_dependency("neural_tangents", descr="ntk")
+
+
+@partial(
```
Comment (PhilipVinc) [inline netket/experimental/driver/vmc_srt_ntk.py:39]: Please add an outermost decorator `@timing.timed` so it shows up as a line.

**Review by PhilipVinc — COMMENTED** [review]
A few quick comments. I'll give a better look later in the week.

This should be imported in the driver init file?

This should also show up in the documentation (add it to the experimental.md file) and you should add a comment in the docstring of this and VMC SRT about their relationship.

This also needs tests. It should be checked that it gives the same results as vmc/srt.
It should be straightforward but you could also coordinate with @lgravina1997 to push the ones he already wrote. 

Also, do note that this driver will probably not pass the tests on the MPI side because you do not gather on the mpi end, unless I'm mistaken, but only with sharding. 
If it is the case, you should raise an error when mpi is in use if you are not implementing it.

As a side note, it would be much nicer to have a single vmc_srt driver with a flag to switch between the two implementations...

File: netket/experimental/driver/vmc_srt.py [hunk 2]
```diff
@@ -32,85 +32,169 @@
 from netket.utils.types import ScalarOrSchedule, Optimizer, PyTree
 from netket.vqs import MCState
 
-from jax.flatten_util import ravel_pytree
+from netket.utils.optional_deps import import_optional_dependency
+
+from netket.jax.sharding import shard_along_axis, all_gather
+
+nt = import_optional_dependency("neural_tangents", descr="neural_tangents")
+einops = import_optional_dependency("einops", descr="einops")
```
Comment (PhilipVinc) [inline netket/experimental/driver/vmc_srt.py:40]: This cannot be here. It must be inside a function.
If it's here, it will be required to load netket.

Comment (PhilipVinc) [reply netket/experimental/driver/vmc_srt.py:40]: I'm ok with adding einops as a dependency of netket, though. So just import it normally and add It as an official dependency

**Comment by gcarleo** [comment]
we are democratizing the access to netket by making it inaccessible to the largest amount of people possible

**Comment by llviteritti** [comment]
In the implementation with the ntk, only one function from neural_tangents is used, which can probably be isolated from the rest. So if we can extract this function we don't need to install the package. @lgravina1997

**Comment by gcarleo** [comment]
this is an exquisite idea

**Comment by PhilipVinc** [comment]
Yes, it's a good idea.
If you want, a lazy way to do it may be to use this script I wrote a while ago

https://gist.github.com/PhilipVinc/1943429d0f7ff44aaea9f04d8dc90124

commit this script and put it in something like `external/` directory, and then run it (paths might be wrong and need changing a bit). 
It will only keep the relevant parts of neural tangents, but allow for future updates if need ever comes.

**Comment by gcarleo** [comment]
what if we just copy those functions inside netket ? does this pose a problem with licenses ?

**Comment by PhilipVinc** [comment]
it's apache 2, so it's ok. the script above does the good thing keeping the original license file

**Comment by lgravina1997** [comment]
I really like the new implementation, its basically the same as we have here and from rough tests it performs the same too. 

One difference I note with the code we use is modularity. 

So with this implementation here we have the vmc_srt function which allows doing VMC while constructing  the NTK "on the fly". 
Then we have nk.vmc that when used with a preconditioner allows doing vmc with normal SR. And we remove the possibility of doing vmc_srt with the full jacobian. 

So the implementations of SR and SRT are fundamentally different and the SRT code is restricted to VMC so that if I wanted to use it e.g with the infidelity I would have to write a new driver from scratch.

What we did instead is have a base class handling ngd optimizations for arbitrary kernels to support both infidelity and vmc, and whatever one might want to add in the future. In this ngd class we support simultaneously SR, SRT with the construction of the operators either from the full jacobian or "on the fly".

The underlying code that you proposed remains the same but it gets structured differently to allow for this generality. So that if we change something it doesnt have to be changed on several drivers and the implementations of SR and SRt are virtually identical. If you are interested I could add the code needed to achieve this. @riccardo-rende @llviteritti @alleSini99 @gcarleo what do you think?

**Comment by gcarleo** [comment]
yes I also don't like the proliferation of drivers, if we could try to merge some it would be highly appreciated!

**Comment by alleSini99** [comment]
Ok for me

**Comment by PhilipVinc** [comment]
A possible signature for the constructor of `VMC_NGD` I like is the following:
```python
class VMC_NGD:
    def __init__(
        self,
        hamiltonian: AbstractOperator,
        optimizer: float | Optimizer, # if float, that's the learning rate, otherwise an optax optimizer
        *,
        # standard stuff
        diag_shift: ScalarOrSchedule,
        linear_solver: Callable[[Array, Array], Array] = cholesky,
        # Select implementation
        mode : JacobianMode = "complex", # could also be real
        use_ntk : bool | Literal["auto"] = "auto" # True/False/auto to switch between SR/minSR (note here ntk has nothing to do with neural tangents)
        onthefly: bool = True # if True + SR => use QGTOnTheFly, If False +SR => QGTJacobian, # If False + ntk => SRt_ntk current implementation, # If True + ntk => new neuraltangents implementation
        # variational state stuff
        variational_state: MCState = None,
        chunk_size : Optional[int] = None, # defaults to variational_state.chunk_size,
        chunk_size_bwd: Optional[int] = None, # defaults to chunk size. 
    ):
```

The only thing I do not love is that one still needs to define `optimizer` , while defining `learning_rate` would be more elegant, but we want to keep allowing users to use a learning rate.

All the logic is already around. We just need a little bit of scaffolding to be able to switch from one to another.
I think this would be a good way to unify the 2 different (or more) interfaces.

**Comment by PhilipVinc** [comment]
For the scope of this PR you could also just implement the case `use_ntk = True` and leave the rest for the future.
I would just ask you do not throw away the current code based on the jacobian, because it has use cases.

**Comment by llviteritti** [comment]
I agree with @lgravina1997  that having a unified driver for everything could be useful. However, since SR is the most widely used optimization method, I would keep it separate from the other drivers. This would allow for independent modifications and improvements in the future without affecting the rest. 

Regarding the previous implementation without NTK, I don’t see a strong reason to keep it. Maintaining both versions would add unnecessary complexity for users, forcing them to choose between implementations in different scenarios. Additionally, I have never encountered a case where the non-NTK version proved more useful. For these reasons, I would proceed with the NTK-based implementation as the sole option. @PhilipVinc

**Comment by lgravina1997** [comment]
@llviteritti consider a situation where you need to compute the full jacobian for purposes parallel to the optimization you are doing. Then its convenient to have it some times. In any case it is almost identical to the SR implementation so supporting it is immediate.

**Comment by PhilipVinc** [comment]
> Regarding the previous implementation without NTK, I don’t see a strong reason to keep it. Maintaining both versions would add unnecessary complexity for users, forcing them to choose between implementations in different scenarios. Additionally, I have never encountered a case where the non-NTK version proved more useful. For these reasons, I would proceed with the NTK-based implementation as the sole option. @PhilipVinc

I'm not suggesting we add more user-facing complexity. A sane default (ntk-onthefly) will be good for 95% of people.
However, it is sometimes handy, when exploring other algorithms, to keep the jacobian around, in which case one would 
resort to the jacobian-based implementation. It makes for a better starting point to implement other algorithms.

The jacobian-based implementation is also easier to understand for most people, so it doubles as good learning material.
